assignment1/fc_net.py

```diff
@@ -29,26 +29,74 @@ def __init__(self, layer_dims, input_dim=3*32*32, num_classes=10,
         
 def initialize_parameters(layer_dims):
     """
-    input: 
+    input:
         an array of the dimensions of each layer in the network (layer 0 is the size of the flattened input, layer L is the output sigmoid)
-    output: 
+    output:
         a dictionary containing the initialized W and b parameters of each layer (W1...WL, b1...bL).
     """
     params = {}
     layer_input_dim = layer_dims[0]
     num_classes = layer_dims[-1]
-    
+
     # input-> hidden_layer_1 -> hidden_layer_2 -> ... -> hidden_layer_last
     for idx, dim in enumerate(layer_dims[1:-1]): # enumrate all hidden layers
         layer_num = str(idx+1)
         params['W' + layer_num] = np.random.randn(layer_input_dim, dim)
         params['b' + layer_num] = np.zeros(dim)
-        layer_input_dim = dim    
-    
+        layer_input_dim = dim
+
     # hidden_layer_last -> output
     num_layers = len(layer_dims) - 1
     params['W' + str(num_layers)] = np.random.randn(layer_input_dim, num_classes)
     params['b' + str(num_layers)] = np.zeros(num_classes)
-    
+
     return params
-    
\ No newline at end of file
+
+def L_model_forward(X, parameters, use_batchnorm):
+    """
+    forward propagation for the [LINEAR->RELU]*(L-1)->LINEAR->SIGMOID computation
+
+    :param X: the data, numpy array of shape (input size, number of examples)
+    :param parameters: the initialized W and b parameters of each layer
+    :param use_batchnorm: a boolean flag used to determine whether to apply batchnorm after the activation
+    :return: (the last post-activation value , a list of all the cache objects)
+    """
+
+    # calculate the number of layers
+    num_layers = len([key for key in d.keys() if key.startswith('W')])
+
+    layer_input = X
+    caches = []
+    for layer_idx in range(num_layers - 1):
+        layer_input, cache = linear_activation_forward(layer_input, parameters['W' + str(layer_idx)], parameters['b' + str(layer_idx)], 'relu')
+        if use_batchnorm:
+            layer_input = apply_batchnorm(layer_input)
+        caches.append(cache)
+
+    # last layer
+    last_post_activation, cache = linear_activation_forward(layer_input, parameters['W' + str(num_layers - 1)], parameters['b' + str(num_layers - 1)], 'sigmoid')
+    caches.append(cache)
```

Comment:
refactor caches as dict of layer_num -> layer_cache
then on backward pass we will use it as caches[layer_num] insead of enumerate

Reply:
The output should be a list as given in the assignment - but i agree its better this way - changing